firebase-firestore/src/main/java/com/google/firebase/firestore/auth/FirebaseAppCheckTokenProvider.java

```diff
@@ -0,0 +1,104 @@
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.appcheck.AppCheckTokenResult;
+import com.google.firebase.appcheck.interop.InternalAppCheckTokenProvider;
+import com.google.firebase.firestore.util.Logger;
+import com.google.firebase.inject.Deferred;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/** FirebaseAppCheckTokenProvider uses Firebase AppCheck to get an AppCheck token. */
+public final class FirebaseAppCheckTokenProvider extends AppCheckTokenProvider {
+
+  private static final String LOG_TAG = "FirebaseAppCheckTokenProvider";
+
+  private static final int MAXIMUM_TOKEN_WAIT_TIME_MS = 30000;
+
+  /** The AppCheck token string. May be null if the token has not been retrieved yet. */
+  @Nullable private String appCheckToken;
+
+  /**
+   * The Task for retrieving the AppCheck token. May be null if the "AppCheck" module is not
+   * available.
+   */
+  @Nullable private Task<String> getTokenTask;
+
+  /** Creates a new FirebaseAppCheckTokenProvider. */
+  public FirebaseAppCheckTokenProvider(
+      Deferred<InternalAppCheckTokenProvider> deferredAppCheckTokenProvider) {
+    if (deferredAppCheckTokenProvider == null) {
+      return;
+    }
+
+    deferredAppCheckTokenProvider.whenAvailable(
+        provider -> {
+          InternalAppCheckTokenProvider internalAppCheckTokenProvider = provider.get();
+          if (internalAppCheckTokenProvider == null) {
+            return;
+          }
+          Task<AppCheckTokenResult> pendingResult =
+              internalAppCheckTokenProvider.getToken(/* forceRefresh= */ false);
+          if (pendingResult == null) {
+            return;
+          }
+          TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+          getTokenTask = taskCompletionSource.getTask();
+          pendingResult.addOnCompleteListener(
+              task -> {
+                if (task.isSuccessful()) {
+                  AppCheckTokenResult result = task.getResult();
+                  if (result.getError() != null) {
+                    Logger.warn(
+                        LOG_TAG,
+                        "Error getting App Check token; using placeholder token instead. Error: "
+                            + result.getError());
+                  }
+                  appCheckToken = result.getToken();
+                  taskCompletionSource.setResult(result.getToken());
+                } else {
+                  Logger.warn(
+                      LOG_TAG, "Unexpected error getting App Check token: " + task.getException());
+                  appCheckToken = null;
+                }
+              });
+        });
+  }
+
+  @Nullable
+  public String getCurrentAppCheckToken() {
+    if (appCheckToken != null) {
+      return appCheckToken;
+    }
+
+    if (getTokenTask == null) {
+      return null;
+    }
```

Comment:
These should not be needed. Note that calling `Tasks.await()` on a completed task has almost no overhead - https://source.corp.google.com/piper///depot/google3/java/com/google/android/gmscore/integ/client/tasks/src/com/google/android/gms/tasks/Tasks.java;l=198

Reply:
Done.